File: ctci/1 Arrays and Strings/1_8.py
# Zero Matrix: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and 
# column are set to 0. 
t2 = [[1, 4, 5, 1],     [-5, 8, 0, 2],    [10, 3, -1, 9]]
def zero_matrix(matrix):
    rows = [False for x in matrix]
    cols = [False for y in matrix[0]]
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            if matrix[i][j] == 0:
                rows[i] = True
                cols[j] = True
    for row_ind, row_bool in enumerate(rows):
        if row_bool:
            matrix[row_ind] = [0 for x in matrix[row_ind]]
        for col_ind,col_bool in enumerate(cols):
            if col_bool:
                matrix[row_ind][col_ind] = 0
    return matrix

zero_matrix(matrix=t2)

File: ctci/practice/1_8.js
// Zero Matrix: Write an algorithm such that if an element in an MxN matrix is 0, its entire row and 
// column are set to 0. 
// Hints:#17, #74, #702

const mat = [[1,2,4,6],
            [3,6,3,6],
            [5,7,5,7],
            [7,0,7,9]]

const zeroMatrix = (matrix) => {
    let row = new Array(matrix.length).fill(false)
    let col = new Array(matrix[0].length).fill(false)

    for(let r in row){
        for(let c in col){
            if(matrix[r][c] === 0){
                row[r] = true
                col[c] = true
            }
        }
    }

    for(let r=0;r < row.length; r++){
        if(row[r]){
            matrix[r] = new Array(matrix[r].length).fill(0)
        }
    }

    for(let c=0;c<col.length;c++){
        if(col[c]){
            for(let nullRow in matrix){
                matrix[nullRow][c] = 0
            }
        }
    }   
}

zeroMatrix(mat)